keps/sig-network/0029-nodelocal-dns-cache.md

```diff
@@ -0,0 +1,145 @@
+
+#### iptables NOTRACK
+
+NOTRACK rules are added for connections to and from the nodelocal dns service ip. Additional rules in FILTER table to whitelist these connections, since the INPUT and OUTPUT chains have a default DROP policy.
+
+A dnscache nanny, similar to the [dnsmasq nanny](https://github.com/kubernetes/dns/tree/master/pkg/dnsmasq) will create the dummy interface and iptables rules. The nanny gets the nodelocal dns service ip address by querying kube-dns for the service name. The nanny will also periodically ensure that the iptables rules are present. The resource usage for periodic iptables check needs to be measured. We can reduce the frequency of the checks by having the nanny query the dns cache and checking/adding rules only if the query fails.
+
+
+[Proposal presentation](https://docs.google.com/presentation/d/1c43cZqbVhGAlw3dSNQIOGuvQmDfKaA2yiAPRoYpa6iY), also shared at the sig-networking meeting on 2018-10-04
+
+Slide 5 has a diagram showing how the new dns cache fits in.
+
+#### Choice of caching agent
+
+The current plan is to run [Unbound dns server](https://www.nlnetlabs.nl/projects/unbound/about/) by default, based on these benchmark[ tests](https://github.com/kubernetes/perf-tests/tree/master/dns).
+
+Will rerun benchmark tests with coreDNS after SO_REUSEPORT support and measure QPS.
+
+Tests were run on a 1.9.7 cluster with 2 nodes, using Unbound 1.7.3 and coreDNS 1.2. Unbound QPS was 3600(caching)/3400(no caching), coreDNS 950(caching)/500(no caching)
+
+
+Based on the prototype/test results, these are the recommended defaults: 
+CPU request: 50m
+Memory Limit : 25m  
+Resource usage and QPS to be recomputed after metrics collection has been added.
+
+#### Metrics
+
+Unbound metrics can be extracted from the program unbound-control. This can be run from the nanny container.
+Unbound-control-setup is required to setup keys/certs for SSL.
+
+
+### Risks and Mitigations
+
+* Having the pods query the nodelocal cache introduces a single point of failure.   
+This is mitigated by assigning a PriorityClass of "system-node-critical", so it is always running. The nanny process will periodically check if the dns agent is running.
```

Comment:
We should talk about how upgrades of the cache will work and what the implications are for outages.  E.g. during a rolling update of the cache queries can be lost and since they are UDP, will be subject to long timeouts.

How will we mitigate this?  

* Consider changing the default timeout paramt from 5s to 0.5s ? Or less.
* Can we do some other tricks to do hitless upgrades?
* Do we strongly recommend draining a node before upgrading this daemonset?

Reply:
That's a good point. 

I am worried setting the timeout too low might cause actual dns queries that are slow to never succeed, if there is a custom upstream server configured.
Can the recommendation be to use custom dns config and point at cluster dns ip for all clusterFIrst mode pods, before upgrading this daemonset?

Reply:
That's a pretty heavy-handed approach - first mutate all of your daemonsets, then update DNS, then mutate them all back.  It would be simpler just to evict.

We could configure clients to have 2 nameservers, and fall back on the cluster-scope server, but clients are unpredictable and some are dumb.  Still might be OK - worth trying, perhaps?  Or understanding more.

Reply:
In the first version of the KEP, I added the option of specifying 2 upstream servers in pods resolv.conf, first one the nodelocal ip followed by cluster dns ip. However, as other reviewers pointed out, only glibc implementation queries the next server if the first one does not respond. musl implementation queries them all in parallel.

One idea is to create the nodelocaldns service with the same selectors as kube-dns/coredns service. That way, it will have the same endpoints. Reaching out to nodelocal ip will hit the kube-dns/coredns pods, as long as the nanny container does not NOTRACK rules. Once nanny container is up and adds the NOTRACK rules and filter table rules, traffic will reach the local instance. If nanny process exits, dns will continue to work. Does this sound like a reasonable approach?

Reply:
Need to verify if this will work in ipvs mode as well.

Reply:
@thockin putting both in the resolv.conf will not work well for alpine images, it will end up hitting the cluster DNS and defeat the purpose. @prameshj that sounds like a viable option, good idea. It may be a little confusing to see the endpoints in the API server but have the traffic stolen by the new iptables rules, but it does solve the HA problem pretty cleanly.

Reply:
It doesn't solve HA entirely.  It solves it in the case of a clean upgrade, but not in the case of a crash.  I am not sure there's a great answer for that, really.  We could make the nanny add/remove NOTRACK when it loses connection to the DNS container, which might cover some of the issues.

Does openshift handle this case?

Reply:
Is there an update planned to handle HA and upgrade issue in this doc ?